File: ci/lint_changelog.py
#!/usr/bin/env python3
"""
Changelog lint: ensure CHANGELOG.md corresponds to the current project version.

Ensures:
- The changelog must contain a release section for the current version
- An [Unreleased] section must exist
- [Unreleased] must appear before the current release section
- Footer links must include entries for both [Unreleased] and the current version

This intentionally does NOT validate changelog contents or section structure.
"""

import re
import sys
from pathlib import Path

from read_version import MetadataError, read_version

ROOT = Path(__file__).resolve().parents[1]
CHANGELOG_FILE = ROOT / "CHANGELOG.md"
METADATA_FILE = ROOT / "metadata.toml"


class LintError(Exception):
    """Raised when the changelog is invalid."""

    pass


def parse_changelog(text: str):
    """
    Extract only the structural elements we care about:
    - section headers with their line numbers
    - footer reference definitions
    """

    unreleased_line = None
    release_lines = {}  # version -> line number
    footer_links = set()

    header_re = re.compile(r"^## \[(?P<name>[^\]]+)]")
    footer_re = re.compile(r"^\[(?P<name>[^\]]+)]:")

    for idx, raw in enumerate(text.splitlines(), start=1):
        line = raw.rstrip()
        if m := header_re.match(line):
            name = m.group("name")
            if name == "Unreleased":
                unreleased_line = idx
            else:
                # Accept any version-looking string here; exact match later
                release_lines[name] = idx
            continue

        if m := footer_re.match(line):
            footer_links.add(m.group("name"))

    return unreleased_line, release_lines, footer_links


def lint(version: str, changelog_file: Path) -> None:
    if not changelog_file.exists():
        raise LintError(f"{changelog_file.name} not found")

    text = changelog_file.read_text()
    if "\r\n" in text:
        raise LintError(
            f"{changelog_file.name} contains CRLF line endings; please convert to LF"
        )

    unreleased_line, releases, footer_links = parse_changelog(text)

    # Must have a release section matching the current version
    if version not in releases:
        raise LintError(
            f"Changelog has no release section for version {version} "
            "(did you forget to run patchChangelog?)"
        )

    # Must have an Unreleased section
    if unreleased_line is None:
        raise LintError("Changelog is missing an [Unreleased] section")

    # Unreleased must appear before the current release
    if unreleased_line > releases[version]:
        raise LintError(
            "[Unreleased] section must appear before the current release section"
        )

    # Footer links must exist
    if "Unreleased" not in footer_links:
        raise LintError("Missing footer link for [Unreleased]")

    if version not in footer_links:
        raise LintError(f"Missing footer link for version {version}")

    # All good
    print(f"Changelog OK for version {version}")


def main() -> None:
    try:
        version = read_version(METADATA_FILE)
        lint(version, CHANGELOG_FILE)
    except LintError as e:
        print("Changelog lint failed:\n", file=sys.stderr)
        print(f"- {e}", file=sys.stderr)
        sys.exit(1)
    except MetadataError as e:
        print("Changelog lint failed to read metadata:\n", file=sys.stderr)
        print(f"- {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: ci/src/lint_changelog.ts
import fs from "node:fs";
import path from "node:path";
import { CHANGELOG_FILE } from "./project_files.ts";
import { readVersion, MetadataError } from "./read_version.ts";

export class LintError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LintError";
  }
}

type ParseResult = {
  unreleasedLine: boolean;
  releases: Map<string, number>;
  oldFormatReleases: Set<string>;
  footerLinks: Set<string>;
};

function parseChangelog(text: string): ParseResult {
  let unreleasedLine: boolean = false;
  const releases = new Map<string, number>();
  const oldFormatReleases = new Set<string>();
  const footerLinks = new Set<string>();

  const headerRe = /^## (?<name>[-.0-9a-zA-Z]+) \((?<date>\d{4}-\d\d-\d\d)\)$/;
  const oldHeaderRe = /^## \[(?<name>[^\]]+)]/;
  const unreleasedRe = /^## \[?unreleased\]?/i;
  const footerRe = /^\[(?<name>[^\]]+)]:/;

  const lines = text.split("\n");

  lines.forEach((raw, idx) => {
    const line = raw.replace(/\r$/, ""); // mimic rstrip for CRLF detection
    const lineNo = idx + 1;

    if (line.match(unreleasedRe)) {
      unreleasedLine = true;
    }

    const headerMatch = line.match(headerRe);
    if (headerMatch?.groups?.name) {
      releases.set(headerMatch.groups.name, lineNo);
    }

    const oldHeaderMatch = line.match(oldHeaderRe);
    if (oldHeaderMatch?.groups?.name) {
      oldFormatReleases.add(oldHeaderMatch.groups.name);
    }

    const footerMatch = line.match(footerRe);
    if (footerMatch?.groups?.name) {
      footerLinks.add(footerMatch.groups.name);
    }
  });

  return { unreleasedLine, releases, oldFormatReleases, footerLinks };
}

export function lint(version: string, changelogFile: string): void {
  if (!fs.existsSync(changelogFile)) {
    throw new LintError(`${path.basename(changelogFile)} not found`);
  }

  const text = fs.readFileSync(changelogFile, "utf8");

  if (text.includes("\r\n")) {
    throw new LintError(
      `${path.basename(
        changelogFile,
      )} contains CRLF line endings; please convert to LF`,
    );
  }

  const { unreleasedLine, releases, oldFormatReleases, footerLinks } =
    parseChangelog(text);

  if (!releases.has(version)) {
    throw new LintError(
      `Changelog has no release section for version ${version} ` +
        "(did you forget to run patchChangelog?)",
    );
  }

  if (unreleasedLine) {
    throw new LintError("Unexpected 'Unreleased' section");
  }

  if (oldFormatReleases.size > 0) {
    throw new LintError(
      `Unexpected releases using the old format (${oldFormatReleases.size})`,
    );
  }

  if (footerLinks.size > 0) {
    throw new LintError(`Unexpected footer links (${footerLinks.size})`);
  }

  console.log(`Changelog OK for version ${version}`);
}

export function main(): void {
  try {
    const version = readVersion();
    lint(version, CHANGELOG_FILE);
  } catch (e) {
    if (e instanceof LintError) {
      console.error("Changelog lint failed:\n");
      console.error(`- ${e.message}`);
      process.exit(1);
    }

    if (e instanceof MetadataError) {
      console.error("Changelog lint failed to read metadata:\n");
      console.error(`- ${e.message}`);
      process.exit(1);
    }

    throw e;
  }
}

if (import.meta.main) main();
